From f4b0a94942062b3b55ef32e2ea20e9a6b3610e96 Mon Sep 17 00:00:00 2001
From: Kaiser <1244713586@qq.com>
Date: Sat, 19 Mar 2022 18:02:21 +0800
Subject: [PATCH] fix brotli link

---
 CMakeLists.txt | 20 +++++++-------------
 1 file changed, 7 insertions(+), 13 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index ecfbb83..236f64d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -32,15 +32,10 @@ if ("${isSystemDir}" STREQUAL "-1")
 endif()
 
 # Find Brotli dependencies
-set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
-find_package(BrotliDec)
-if (NOT BROTLIDEC_FOUND)
-    message(FATAL_ERROR "librotlidec is needed to build woff2.")
-endif ()
-find_package(BrotliEnc)
-if (NOT BROTLIENC_FOUND)
-    message(FATAL_ERROR "librotlienc is needed to build woff2.")
-endif ()
+find_package(PkgConfig REQUIRED)
+pkg_check_modules(libbrotlidec REQUIRED IMPORTED_TARGET libbrotlidec)
+pkg_check_modules(libbrotlienc REQUIRED IMPORTED_TARGET libbrotlienc)
+pkg_check_modules(libbrotlicommon REQUIRED IMPORTED_TARGET libbrotlicommon)
 
 # Set compiler flags
 if (NOT CANONICAL_PREFIXES)
@@ -64,8 +59,7 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${COMMON_FLAG}")
 set(CMAKE_CXX_STANDARD 11)
 
 # Set search path for our private/public headers as well as Brotli headers
-include_directories("src" "include"
-                    "${BROTLIDEC_INCLUDE_DIRS}" "${BROTLIENC_INCLUDE_DIRS}")
+include_directories("src" "include")
 
 # Common part used by decoder and encoder
 add_library(woff2common
@@ -77,7 +71,7 @@ add_library(woff2common
 add_library(woff2dec
             src/woff2_dec.cc
             src/woff2_out.cc)
-target_link_libraries(woff2dec woff2common "${BROTLIDEC_LIBRARIES}")
+target_link_libraries(woff2dec woff2common PkgConfig::libbrotlidec PkgConfig::libbrotlicommon)
 add_executable(woff2_decompress src/woff2_decompress.cc)
 target_link_libraries(woff2_decompress woff2dec)
 
@@ -88,7 +82,7 @@ add_library(woff2enc
             src/normalize.cc
             src/transform.cc
             src/woff2_enc.cc)
-target_link_libraries(woff2enc woff2common "${BROTLIENC_LIBRARIES}")
+target_link_libraries(woff2enc woff2common PkgConfig::libbrotlienc PkgConfig::libbrotlicommon)
 add_executable(woff2_compress src/woff2_compress.cc)
 target_link_libraries(woff2_compress woff2enc)
 
-- 
2.32.0

