From faf787ebe2fd2c5bdd94f84259b7913c120aee56 Mon Sep 17 00:00:00 2001
From: Kaiser <kaiserlancelot123@gmail.com>
Date: Fri, 15 Apr 2022 19:28:07 +0800
Subject: [PATCH] fix link error

---
 gio/gthreadedresolver.c | 29 +++++++++++++++++++++++++----
 1 file changed, 25 insertions(+), 4 deletions(-)

diff --git a/gio/gthreadedresolver.c b/gio/gthreadedresolver.c
index aeeb40e..a695324 100644
--- a/gio/gthreadedresolver.c
+++ b/gio/gthreadedresolver.c
@@ -39,6 +39,27 @@
 
 G_DEFINE_TYPE (GThreadedResolver, g_threaded_resolver, G_TYPE_RESOLVER)
 
+static void
+bail (void)
+{
+  fprintf (stderr, "In order to solve the link error, it is assumed that DNS is not used, and it is terminated directly\n");
+  abort ();
+}
+
+static int
+k_dn_expand (const u_char *, const u_char *, const u_char *, char *, int)
+{
+  bail ();
+  return 0;
+}
+
+static int
+k_res_nquery (struct __res_state*, const char *, int, int, unsigned char *, int)
+{
+  bail ();
+  return 0;
+}
+
 static void
 g_threaded_resolver_init (GThreadedResolver *gtr)
 {
@@ -542,7 +563,7 @@ expand_name (const gchar   *rrname,
 {
   int expand_result;
 
-  expand_result = dn_expand (answer, end, *p, namebuf, namebuf_len);
+  expand_result = k_dn_expand (answer, end, *p, namebuf, namebuf_len);
   if (expand_result < 0 || end - *p < expand_result)
     {
       g_set_error (error, G_RESOLVER_ERROR, G_RESOLVER_ERROR_INTERNAL,
@@ -793,7 +814,7 @@ g_resolver_records_from_res_query (const gchar      *rrname,
     {
       int expand_result;
 
-      expand_result = dn_expand (answer, end, p, namebuf, sizeof (namebuf));
+      expand_result = k_dn_expand (answer, end, p, namebuf, sizeof (namebuf));
       if (expand_result < 0 || end - p < expand_result + 4)
         {
           /* Not possible to recover parsing as the length of the rest of the
@@ -818,7 +839,7 @@ g_resolver_records_from_res_query (const gchar      *rrname,
     {
       int expand_result;
 
-      expand_result = dn_expand (answer, end, p, namebuf, sizeof (namebuf));
+      expand_result = k_dn_expand (answer, end, p, namebuf, sizeof (namebuf));
       if (expand_result < 0 || end - p < expand_result + 10)
         {
           /* Not possible to recover parsing as the length of the rest of the
@@ -1119,7 +1140,7 @@ do_lookup_records (GTask         *task,
     {
       g_byte_array_set_size (answer, len * 2);
 #if defined(HAVE_RES_NQUERY)
-      len = res_nquery (&res, lrd->rrname, C_IN, rrtype, answer->data, answer->len);
+      len = k_res_nquery (&res, lrd->rrname, C_IN, rrtype, answer->data, answer->len);
 #else
       len = res_query (lrd->rrname, C_IN, rrtype, answer->data, answer->len);
 #endif
-- 
2.32.0

