From 6ed9c17adbf03b60c5d6bdcf4df4da4991b282af Mon Sep 17 00:00:00 2001
From: Kaiser <1244713586@qq.com>
Date: Sat, 19 Mar 2022 18:18:47 +0800
Subject: [PATCH] remove HAVE_SYMVER_ALIAS_SUPPORT

---
 lib/puny_decode.c | 9 ---------
 lib/puny_encode.c | 9 ---------
 2 files changed, 18 deletions(-)

diff --git a/lib/puny_decode.c b/lib/puny_decode.c
index ee2c534..4d6c596 100644
--- a/lib/puny_decode.c
+++ b/lib/puny_decode.c
@@ -242,12 +242,3 @@ punycode_decode (size_t input_length,
   /* cannot overflow because out <= old value of *output_length */
   return punycode_success;
 }
-
-/* Create a compatibility symbol if supported.  Hidden references make
-   the target symbol hidden, hence the alias.  */
-#ifdef HAVE_SYMVER_ALIAS_SUPPORT
-__typeof__ (_idn2_punycode_decode_internal) _idn2_punycode_decode
-  __attribute__((visibility ("default"),
-		 alias ("_idn2_punycode_decode_internal")));
-__asm__ (".symver _idn2_punycode_decode, _idn2_punycode_decode@IDN2_0.0.0");
-#endif
diff --git a/lib/puny_encode.c b/lib/puny_encode.c
index 909047b..aae4226 100644
--- a/lib/puny_encode.c
+++ b/lib/puny_encode.c
@@ -243,12 +243,3 @@ punycode_encode (size_t input_length_orig,
   *output_length = out;
   return punycode_success;
 }
-
-/* Create a compatibility symbol if supported.  Hidden references make
-   the target symbol hidden, hence the alias.  */
-#ifdef HAVE_SYMVER_ALIAS_SUPPORT
-__typeof__ (_idn2_punycode_encode_internal) _idn2_punycode_encode
-  __attribute__((visibility ("default"),
-		 alias ("_idn2_punycode_encode_internal")));
-__asm__ (".symver _idn2_punycode_encode, _idn2_punycode_encode@IDN2_0.0.0");
-#endif
-- 
2.32.0

